edb/schema/rewrites.py

```diff
@@ -0,0 +1,301 @@
+        self,
+        schema: s_schema.Schema,
+        context: sd.CommandContext,
+        field: so.Field[Any],
+        value: s_expr.Expression,
+        track_schema_ref_exprs: bool = False,
+    ) -> s_expr.CompiledExpression:
+        if field.name == 'expr':
+            from edb.ir import pathid
+
+            parent_ctx = self.get_referrer_context_or_die(context)
+            source = parent_ctx.op.get_object(schema, context)
+            assert isinstance(source, s_types.Type)
+            # XXX: in_ddl_context_name is disabled for now because
+            # it causes the compiler to reject DML; we might actually
+            # want it for something, though, so we might need to
+            # improve that restriction.
+            # parent_vname = source.get_verbosename(schema)
+            # pol_name = self.get_verbosename(parent=parent_vname)
+            # in_ddl_context_name = pol_name
+
+            kind = self._get_kind(schema)
+
+            anchors = {}
+            if qltypes.RewriteKind.Update == kind:
+                anchors['__old__'] = pathid.PathId.from_type(
+                    schema,
+                    source,
+                    typename=sn.QualName(module='__derived__', name='__old__'),
+                )
+
+            anchors['__specified__'] = pathid.PathId.from_type(
+                schema,
+                source,
+                typename=sn.QualName(
+                    module='__derived__', name='__specified__'
+                ),
+            )
+
+            singletons = frozenset(anchors.values())
+
+            assert isinstance(source, s_types.Type)
+
+            return type(value).compiled(
+                value,
+                schema=schema,
+                options=qlcompiler.CompilerOptions(
+                    modaliases=context.modaliases,
+                    schema_object_context=self.get_schema_metaclass(),
+                    anchors=anchors,
+                    singletons=singletons,
+                    apply_query_rewrites=not context.stdmode,
+                    track_schema_ref_exprs=track_schema_ref_exprs,
+                    # in_ddl_context_name=in_ddl_context_name,
+                    detached=True,
+                ),
+            )
+        else:
+            return super().compile_expr_field(
+                schema, context, field, value, track_schema_ref_exprs
+            )
+
+    def get_dummy_expr_field_value(
+        self,
+        schema: s_schema.Schema,
+        context: sd.CommandContext,
+        field: so.Field[Any],
+        value: Any,
+    ) -> Optional[s_expr.Expression]:
+        if field.name == 'expr':
+            return s_expr.Expression(text='false')
+        else:
+            raise NotImplementedError(f'unhandled field {field.name!r}')
+
+    def validate_object(
+        self,
+        schema: s_schema.Schema,
+        context: sd.CommandContext,
+    ) -> None:
+        # XXX: verify we don't have the same bug as access policies
```

Comment:
Oh yeah we'll need to think about linkprop interactions. Probably these won't work on linkprops unless we change a bunch about how we compile them